### PR TITLE
arch/x86_64: correctly compute xray instrumentation addresses

### LINKED ISSUE #1265: Google XRay instrumentation don't seem to work
Hi, I'm trying to get `uftrace` working with Google XRay dynamic instrumentation. When patching all functions `-P .` on my own binary, none get traced in the final output.

I decided to run the unit tests (`make test`). Sure enought, test 140 `dynamic_xray` fails. Here's a sample log
```
build command: clang -o t-abc -fno-inline -fno-builtin -fno-ipa-cp -fno-omit-frame-pointer -D_FORTIFY_SOURCE=0  -fxray-instrument -fxray-instruction-threshold=1  s-abc.c  -lstdc++ 
test command: <redacted>/uftrace live --no-pager --no-event -L<redacted> -P 'a.?' t-abc
=========== original ===========
# DURATION     TID     FUNCTION
   1.002 us [ 70346] | getpid();

===========  result  ===========

=========== expected ===========
main() {
   a() {
     getpid();
   } /* a */
 } /* main */
140 dynamic_xray        : NG NG NG NG NG NG NG NG NG NG
```

I ran `gdb` and I found out that this [check](https://github.com/namhyung/uftrace/blob/master/arch/x86_64/mcount-dynamic.c#L352) always prevent the instrumentation of the functions.

I'm not entirely familiar with Google XRay (raw technicals details about binary section are scare IMO). If you have any pointers/ideas/things I could check, I wouldn't mind  digging deeper into the problem and submitting a PR.

Relevant information:
```
$ uname -a # (ArchLinux)
Linux <redacted> 5.12.4-arch1-2 #1 SMP PREEMPT Sat, 15 May 2021 20:58:02 +0000 x86_64 GNU/Linux

$ clang -v
clang version 11.1.0
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-pc-linux-gnu/11.1.0
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-pc-linux-gnu/8.4.0
Found candidate GCC installation: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/11.1.0
Found candidate GCC installation: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/8.4.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-pc-linux-gnu/8.4.0
Found candidate GCC installation: /usr/lib64/gcc/x86_64-pc-linux-gnu/11.1.0
Found candidate GCC installation: /usr/lib64/gcc/x86_64-pc-linux-gnu/8.4.0
Selected GCC installation: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/11.1.0
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Selected multilib: .;@m64

$ uftrace --version # (actualy master branch)
uftrace v0.9.4-239-g77c01193 ( x86_64 dwarf python luajit tui perf sched dynamic )

```

### PR DESCRIPTION
**Pull request by gpollo**
Fixes #1265.

See the commit message for more technical informations. In short, I believe that the `xray_instr_map` section may have changed a bit in how its content is referenced. I haven't look at any LLVM/clang code to confirm that.

Test 140 `dynamic_xray` should also pass now.

~~NOTE: this has not been tested for non-PIE executable. I can't get them to compile on my system, thus I can't test them.~~ It works for non-PIE executable also.

### CODE REVIEW

**Comment by namhyung** [comment]
Thanks for the analysis and the fix.  Let me check the changes in the LLVM/X-ray (sorry, probably it'd take some time...)

**Comment by gpollo** [comment]
I was able to test for non-PIE executable (I thought the `clang` options was `fno-pie`, but its actually `-nopie`). The way the entries are put in `xray_instr_map` are the same, thus the same correction need to be applied whether the binary is PIE or non-PIE.

File: arch/x86_64/mcount-dynamic.c [hunk 1]
```diff
@@ -24,10 +24,10 @@ extern void __xray_entry(void);
...
 
 struct xray_instr_map {
-	unsigned long addr;
-	unsigned long entry;
-	unsigned long type;
-	unsigned long count;
+	uint64_t instr_addr;
+	uint64_t func_begin_addr;
+	uint8_t type;
+	uint8_t padding[15];
```
Comment (namhyung) [inline arch/x86_64/mcount-dynamic.c:30]: Please update this part to be in sync with the LLVM; https://github.com/llvm/llvm-project/blob/d480f968ad8b56d3ee4a6b6df5532d485b0ad01e/compiler-rt/lib/xray/xray_interface_internal.h#L24

Also I'd like to avoid unnecessary changes due to renaming if possible.

File: arch/x86_64/mcount-dynamic.c [hunk 2]
```diff
@@ -141,17 +143,11 @@ static void read_xray_map(struct arch_dynamic_info *adi,
...
-		for (i = 0; i < adi->xrmap_count; i++) {
-			xrmap = &adi->xrmap[i];
+	for (i = 0; i < adi->xrmap_count; i++) {
+		struct xray_instr_map *xrmap = &adi->xrmap[i];
 
-			xrmap->addr  += offset;
-			xrmap->entry += offset;
-		}
+		xrmap->instr_addr += offset + (shdr->sh_offset + i * sizeof(struct xray_instr_map));
+		xrmap->func_begin_addr += offset + (shdr->sh_offset + i * sizeof(struct xray_instr_map) + 8);
```
Comment (namhyung) [inline arch/x86_64/mcount-dynamic.c:150]: It seems this logic should be applied only when the X-ray sled entry version is 2 or above.

**Review by namhyung — COMMENTED** [review]
Thanks for this work.  I didn't try it for a while.  There seems to be many changes in this area, and the third byte after two addesses in the struct looks like a version number.  And we should check it when updating the addresses.

**Comment by honggyukim** [comment]
Thanks for doing this. I see that this fixes the current problem functionally.

> XRay sleds may now use PC-relative addressing, see llvm-project
commit 10bc12588 for more information.

But could you add a web link for the `commit 10bc12588` in the commit message?

**Comment by honggyukim** [comment]
Please add `Signed-off-by` line in the commit message as well.

File: arch/x86_64/mcount-dynamic.c [hunk 3]
```diff
@@ -141,16 +146,16 @@ static void read_xray_map(struct arch_dynamic_info *adi,
...
-		for (i = 0; i < adi->xrmap_count; i++) {
-			xrmap = &adi->xrmap[i];
+	for (i = 0; i < adi->xrmap_count; i++) {
+		xrmap = &adi->xrmap[i];
 
-			xrmap->addr  += offset;
-			xrmap->entry += offset;
+		if (xrmap->version == 2) {
+			xrmap->address += offset + (shdr->sh_offset + i * sizeof(struct xray_instr_map));
+			xrmap->function += offset + (shdr->sh_offset + i * sizeof(struct xray_instr_map) + 8);
```
Comment (namhyung) [inline arch/x86_64/mcount-dynamic.c:154]: I prefer passing the variable to the sizeof like `sizeof(*xrmap)` as it's shorter and easier to deal with renames.

**Review by namhyung — APPROVED** [review]
LGTM!
